### PR TITLE
Implement simple sphinx.yaml config parsing

### PR DESCRIPTION
**Pull request by chrisjsewell**
Subject: A simple implementation of #9040

### Feature or Bugfix

- Feature

### CODE REVIEW

File: sphinx/config.py
Comment (tk0miya) [inline sphinx/config.py]: What purpose is the return value of this function? I think the filename (and filetype) of config should be encapsulated to this class. So it would be better to rename this as `Config.exists?(confdir)` instead.

Or it's enough to raise an error on reading instead of checking existence.

File: sphinx/config.py
Comment (tk0miya) [inline sphinx/config.py]: Why this function takes `tags`?

File: tests/test_config.py
Comment (tk0miya) [inline tests/test_config.py]: It would be better to separate the core feature of the config object to another cases (overrides, `in` API, and so on). They're not related at all.

**Review by tk0miya — COMMENTED** [review]
LGTM with nits.

File: sphinx/config.py
Comment (chrisjsewell) [reply sphinx/config.py]: so that they can be added to the namespace before it is returned

File: tests/test_config.py
Comment (chrisjsewell) [reply tests/test_config.py]: is it enough to just remove the overrides and only keep the initial tests?

```
    # simple values
    assert 'project' in cfg.__dict__
    assert cfg.project == 'Sphinx <Tests>'
    assert cfg.templates_path == ['_templates']
```

File: sphinx/application.py
Comment (chrisjsewell) [inline sphinx/application.py]: is it necessary to have this check here, or as you say below, is it's enough to just raise an error on reading (@ line 214), instead of checking existence?

File: sphinx/config.py
Comment (chrisjsewell) [reply sphinx/config.py]: see above 👆

File: sphinx/application.py
Comment (tk0miya) [reply sphinx/application.py]: Good point. It seems nobody touches the config file until L214. So it's enough to raise an exception inside `Config.read()`.

File: sphinx/config.py
Comment (tk0miya) [reply sphinx/config.py]: Hmm? How is it used? AFAIK, nobody will be able to access it. I still don't understand why it's needed. In `conf.py`'s case, it allows to change the configurations in `conf.py` dynamically. It means the tags variable is the input of `conf.py`, not a return value of the `read_config_*()` function.

File: tests/test_config.py
Comment (tk0miya) [reply tests/test_config.py]: Yes :-)

Comment (chrisjsewell) [reply tests/test_config.py]: done 👍

File: sphinx/application.py
Comment (chrisjsewell) [reply sphinx/application.py]: done 👍

File: sphinx/config.py
Comment (chrisjsewell) [reply sphinx/config.py]: ah ok I see (I hadn't actually heard about tags until recently), I've removed it from the function then 👍

File: sphinx/config.py
Comment (tk0miya) [inline sphinx/config.py]: I noticed this error is related to the configuration. So it would be much better to use `ConfigError` instead.

File: sphinx/config.py
Comment (tk0miya) [inline sphinx/config.py]: minor: It would be better to mention the filename `config.yml` in these error messages. It lets users know what file they should fix.

Note: I know this is a trial implementation. So I marked this as minor.

Comment (chrisjsewell) [reply sphinx/config.py]: done and added tests for parsing failures

File: sphinx/config.py
Comment (chrisjsewell) [reply sphinx/config.py]: done

File: tests/test_config.py
Comment (tk0miya) [inline tests/test_config.py]: Why mocking logger?

Comment (chrisjsewell) [reply tests/test_config.py]: just a copy/paste, now removed

File: setup.py
Comment (tk0miya) [inline setup.py:35]: Note: I need to update a dependency list on merging this.

**Comment by tk0miya** [comment]
LGTM. Thank you for your work!

**Comment by chrisjsewell** [comment]
> LGTM. Thank you for your work!

thanks for the quick review!

**Comment by chrisjsewell** [comment]
BTW, what is necessary to get this merged?

**Comment by tk0miya** [comment]
I think the discussion in #9040 has not been agreed yet. So I can't merge this until then.

File: sphinx/config.py
Comment (jakobandersen) [inline sphinx/config.py:174]: I think the initial version should still allow both files, such that one can change to YAML incrementally, and in the end still have dynamic things in ``conf.py`` and static things in ``sphinx.yaml``.

Comment (chrisjsewell) [reply sphinx/config.py:174]: No, I disagree.

1. For a version 1 it is best to keep it simple, strict and not add functionality that may need to be later removed
2. We are not removing conf.py, so there is no need for an incremental transition
3. It somewhat defeats the purpose of a static configuration, if you would also have to read `conf.py` to find all the variables

Comment (choldgraf) [reply sphinx/config.py:174]: and importantly, I think that a decision like this would require more discussion, and isn't strictly a blocker on a first-step implementation, so IMO we should defer that to a future PR after this feature exists and people have had time to use it and get a feel for how it could be improved

**Review by choldgraf — APPROVED** [review]

File: sphinx/config.py
Comment (jakobandersen) [reply sphinx/config.py:174]: Well, if you are sitting with an existing project and want the benefit of the static configuration in the YAML file, you will with this be forced to convert _all_ in one go. There is not incremental path to check that you didn't make an error. And if you have just a single dynamic variable (say ``version``) there is no way to use this.
Regarding 3, as argued in the issue: all projects have conventions in one form or another. One such convention could be to ban ``conf.py`` if you want to guarantee static configuration. Why should that block others from using Sphinx in a different way?

Comment (chrisjsewell) [reply sphinx/config.py:174]: As I and others have already noted, this is a version 1 implementation that implements the minimum desired functionality and (intentionally) does not preclude the later addition of co-file support.
If you greatly desire such functionality, then I would encourage you to open a follow-up PR to achieve this.
For now this allows "basic" users of sphinx to use the sphinx.yaml, and in no way affects your current sphinx workflow

**Comment by chrisjsewell** [comment]
@tk0miya I feel this now has enough tacit approval? https://github.com/sphinx-doc/sphinx/issues/9040#issuecomment-833693871, https://github.com/sphinx-doc/sphinx/issues/9040#issuecomment-833693871, https://github.com/sphinx-doc/sphinx/issues/9040#issuecomment-833910428
(and, as mentioned, allows a route for further non-breaking iteration)

File: doc/usage/quickstart.rst
Comment (chrisjsewell) [inline doc/usage/quickstart.rst:254]: FYI @ericholscher

File: doc/usage/quickstart.rst
Comment (humitos) [inline doc/usage/quickstart.rst]: In my _basic_ English knowledge, I understand that both may or may not co-exist. So, in my head, this means that I can use `conf.py` and `sphinx.yaml` files together or only one of them. I'd change this sentence to "the two _must not_ co-exist" to avoid this confusion.

**Review by humitos — APPROVED** [review]
I'm happy with this PR and I'm 👍🏼 on merging it.

I think in another PR we could improve the documentation to show an example of the `sphinx.yaml` and also guide users with small dynamic requirements about how to proceed (in #9040 @hukkinj1 mentioned `bump2version` for example). We could also point them to write a small extension in case they have bigger dynamic requirements as well.

It may be good to mention also that no dynamic YAML configs are allowed (on purpose) in the `sphinx.yaml` file. In https://github.com/sphinx-doc/sphinx/issues/9040#issuecomment-817009677 I mentioned how to this and I'd expect users wanting this and probably complaining about this not working 😄

File: doc/usage/quickstart.rst
Comment (choldgraf) [reply doc/usage/quickstart.rst]: +1 to "must not" or "cannot" instead of "may not"

Comment (chrisjsewell) [inline doc/usage/quickstart.rst]:
```suggestion
   in-place of the :file:`conf.py` (the two must not co-exist).
```

Comment (chrisjsewell) [reply doc/usage/quickstart.rst]: done 👍

**Pull request closed without merging**
